File: java/app/src/main/java/com/youknow/movie/ui/upcoming/UpcomingPresenter.java
package com.youknow.movie.ui.upcoming;

import android.view.View;

import com.youknow.domain.model.SimpleMovie;
import com.youknow.domain.usecase.GetUpcomingMovies;
import com.youknow.movie.ui.nowplaying.NowPlayingContract.Presenter;

import java.util.List;

public class UpcomingPresenter implements Presenter {

    private UpcomingContract.View view;
    private GetUpcomingMovies getMovies;

    UpcomingPresenter(UpcomingContract.View view, GetUpcomingMovies getMovies) {
        this.view = view;
        this.getMovies = getMovies;
    }

    @Override
    public void getUpcomingMovies() {
        view.showProgressBar(View.VISIBLE);
        view.hideError();

        getMovies.get(new GetUpcomingMovies.GetMoviesCallback() {
            @Override
            public void onMoviesLoaded(List<SimpleMovie> movies) {
                view.showProgressBar(View.GONE);
                view.onMoviesLoaded(movies);
            }
        });
    }
}

File: kotlin-rx-koin/app/src/main/java/com/youknow/movie/ui/upcoming/UpcomingPresenter.kt
package com.youknow.movie.ui.upcoming

import android.view.View
import com.youknow.domain.usecase.GetUpcomingMoviesUsecase
import com.youknow.movie.R
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.error

class UpcomingPresenter(
    private val view: UpcomingContract.View,
    private val getMovies: GetUpcomingMoviesUsecase,
    private val ioScheduler: Scheduler = Schedulers.io(),
    private val uiScheduler: Scheduler = AndroidSchedulers.mainThread(),
    private val disposable: CompositeDisposable = CompositeDisposable()
) : UpcomingContract.Presenter, AnkoLogger {

    override fun unsubscribe() {
        disposable.clear()
    }

    override fun getUpcomingMovies() {
        view.showProgressBar(View.VISIBLE)
        view.hideError()

        disposable.add(
            getMovies.get()
                .subscribeOn(ioScheduler)
                .observeOn(uiScheduler)
                .subscribe({ movies ->
                    view.showProgressBar(View.GONE)
                    if (movies.isNullOrEmpty()) {
                        view.onError(R.string.err_movies_not_exists)
                    } else {
                        view.onMoviesLoaded(movies)
                    }
                }, {t ->
                    view.showProgressBar(View.GONE)
                    view.onError(R.string.err_get_movies_failed)
                    error("[Y.M.] getMoviesNowPlaying - failed: ${t.message}", t)
                })
        )
    }

}